### PR TITLE
n + 1 fxes & views specs

### PR DESCRIPTION
**Pull request by jmonto55**
This pull request contains all the steps needed to fix n + 1 problems and build the view specs for the [blog app](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md) RoR project, including:

- n + 1 fixes in all cases needed.
### Before

before

### After

after

Integration specs for:

- User index page:

    - I can see the username of all other users.
    - I can see the profile picture for each user.
    - I can see the number of posts each user has written.
    - When I click on a user, I am redirected to that user's show page.
- User show page:

    - I can see the user's profile picture.
    - I can see the user's username.
    - I can see the number of posts the user has written.
    - I can see the user's bio.
    - I can see the user's first 3 posts.
    - I can see a button that lets me view all of a user's posts.
    - When I click a user's post, it redirects me to that post's show page.
    - When I click to see all posts, it redirects me to the user's post's index page.
- User post index page:

    - I can see the user's profile picture.
    - I can see the user's username.
    - I can see the number of posts the user has written.
    - I can see a post's title.
    - I can see some of the post's body.
    - I can see the first comments on a post.
    - I can see how many comments a post has.
    - I can see how many likes a post has.
    - I can see a section for pagination if there are more posts than fit on the view.
    - When I click on a post, it redirects me to that post's show page.
- Post show page:

    - I can see the post's title.
    - I can see who wrote the post.
    - I can see how many comments it has.
    - I can see how many likes it has.
    - I can see the post body.
    - I can see the username of each commentor.
    - I can see the comment each commentor left.

### CODE REVIEW

File: spec/system/users/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      visit root_path
+      expect(page).to have_content('Number of posts:')
+    end
+
+    it 'When I click on a user, I am redirected to that user show page' do
+      visit root_path
+      click_link 'Jose'
+      expect(page).to have_content('software engineer')
+    end
+  end
```
Comment (rloterh) [inline spec/system/users/users_index_spec.rb]: - [x] Let's ensure to include test for the below scenarios for _user index_ views: 
   * I can see the username of all other users

File: spec/system/users/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    expect(page).to have_content('Salo')
+  end
+  it 'User should see the see the user\'s bio' do
+    visit user_path(@new_user)
+    expect(page).to have_content('software engineer')
+  end
+  it 'User should see the see the user\'s recent posts' do
+    visit user_path(@new_user)
+    expect(page).to have_content('I feel awesome')
+    expect(page).to have_content('Hello Rails')
```
Comment (rloterh) [inline spec/system/users/users_show_spec.rb]: - [ ] _Line 33_ should rather test for the below scenario for _user show_ views: 
   * I can see the user's first 3 posts

File: spec/system/posts/posts_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        expect(page).to have_content 'testing'
+      end
+
+      it 'I can see the user\'s username.' do
+        expect(page).to have_content 'Jose'
+      end
+
+      it 'I can see how many comments a post has.' do
+        expect(page).to have_content 'Comments: 7'
+      end
```
Comment (rloterh) [inline spec/system/posts/posts_index_spec.rb]: - [x] Let's ensure to include a test for the below scenarios for _post index_ views: 
   * I can see the first comments on a post.

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Jose, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project, but you are almost there! 
🏁 🚦  ⛪ 🚜   🏃 🌴  🏫 🌳 🏘️ 

## Highlights 👍  
- N + 1 problem addressed!
- Professional ReadMe!

## Required Changes ♻️
_Kindly check the inline comments under the review._

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/system/users/users_show_spec.rb (hunk 2 above)
Comment (jmonto55) [reply spec/system/users/users_show_spec.rb]: > * [ ]  _Line 33_ should rather test for the below scenario for _user show_ views:
>   
>   * I can see the user's first 3 posts

Hi @rloterh thank you so much for your review 💯. However, the blog app project does not have a method to show the user's first 3 posts, it is actually quite the opposite, we were supposed to build a method to show the user's last 3 posts. I know the requirement in this specific milestone says "user's first 3 posts" but that method is not part of the User Model and does not exist in the project at all. Below you can see the requirement says to build a custom method for the last 3, not the first 3 👇🏽

review

File: spec/system/users/users_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    visit user_path(@new_user)
+    expect(page).to have_content('software engineer')
+  end
+  it 'should see the user\'s last 3 posts only' do
+    visit user_path(@new_user)
+    expect(page).to have_content('I feel awesome')
+    expect(page).to have_content('Hello Rails')
+    expect(page).to have_content('this is true')
+    expect(page).not_to have_content('this is my first post')
+  end
```
Comment (addod19) [inline spec/system/users/users_show_spec.rb]: - [ ] As mentioned in the previous review, we are required to write a test for user first 3 post, then create a method for the  first 3 post(TDD), meaning, we write test before implementing the actual code
Screenshot from 2023-05-16 18-52-18

File: spec/system/posts/posts_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+RSpec.describe 'Posts index page', type: :system do
+  describe 'check the content of the index page' do
+    before(:each) do
+      @user = User.create(name: 'Jose', photo: 'pic.jpg', bio: 'software developer')
+      @post = Post.create(author: @user, title: 'Tests', text: 'testing', likes_counter: 7, comments_counter: 5)
+      @comment1 = Comment.create(text: 'My first comment', author: @user, post: @post)
+      @comment2 = Comment.create(text: 'My second comment', author: @user, post: @post)
+    end
+    background { visit user_posts_path(@user.id) }
```
Comment (addod19) [inline spec/system/posts/posts_index_spec.rb]: - [ ] As I run the `rspec spec/system` command, I encountered the error as show in the screenshot below. Kindly investigate and fix accordingly
Screenshot from 2023-05-16 18-46-13

**Review by addod19 — CHANGES_REQUESTED** [review]
### Hi TEAM, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- [ ] Nice job addressing some of the comments mentioned in the previous review :rocket: 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by anagudelogu — APPROVED** [review]
# STATUS: APPROVED! 🍾 🎆 🥇 

Hi @jmonto55, 
I'm Andres.

You did a great job working on this milestone, your project is complete! 🚀 
There is nothing else to say other than... it's time to merge it 🤖  
Congratulations! 🎉 



## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Great UI design. 💻 
✅ Descriptive PR title, summary and commit messages. 💬 


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**